source/Ocl/Parsing/OclParser.cs

```diff
@@ -0,0 +1,120 @@
+using System;
+using System.Linq;
+using Sprache;
+
+namespace Octopus.Ocl.Parsing
+{
+    internal static class OclParser
+    {
+        private static readonly Parser<Char> Comma = Parse.Char(',');
+        private static readonly Parser<Char> ArrayOpen = Parse.Char('[');
+        private static readonly Parser<Char> ArrayClose = Parse.Char(']');
+
+        private static readonly Parser<Char> BlockOpen = Parse.Char('{');
+        private static readonly Parser<Char> BlockClose = Parse.Char('}');
+
+        public static Parser<T> SameLineToken<T>(this Parser<T> parser)
+            => from leading in ParserCommon.WhitespaceExceptNewline.Many()
+                from item in parser
+                from trailing in ParserCommon.WhitespaceExceptNewline.Many()
+                select item;
+
+        private static readonly Parser<string> Name =
+            from name in Parse.Char(c => c == '_' || char.IsLetterOrDigit(c), "letter, digit, _")
+                .AtLeastOnce()
+                .Text()
+            select name;
+
+
+        private static readonly Parser<int> IntegerLiteral =
+            from number in Parse.Number
+            select int.Parse(number);
+
+        private static readonly Parser<decimal> DecimalLiteral =
+            from number in Parse.DecimalInvariant
+            select decimal.Parse(number);
+
+        private static readonly Parser<string[]> QuotedStringArrayLiteral =
+            from open in ArrayOpen.Token()
+            from values in QuotedStringParser.QuotedStringLiteral.DelimitedBy(Comma.Token())
+            from close in ArrayClose.Token()
+            select values.ToArray();
+
+        private static readonly Parser<string[]> EmptyArrayLiteral =
+            from open in ArrayOpen.Token()
+            from close in ArrayClose.Token()
+            select new string[0];
+
+        private static readonly Parser<int[]> IntArrayLiteral =
+            from open in ArrayOpen.Token()
+            from values in IntegerLiteral.DelimitedBy(Comma.Token())
+            from close in ArrayClose.Token()
+            select values.ToArray();
+
+        private static readonly Parser<decimal[]> DecimalArrayLiteral =
+            from open in ArrayOpen.Token()
+            from values in DecimalLiteral.DelimitedBy(Comma.Token())
+            from close in ArrayClose.Token()
+            select values.ToArray();
+
+        private static readonly Parser<object> Literal =
+            QuotedStringParser.QuotedStringLiteral
+                .Or<object>(HeredocParser.Literal)
+                .Or(DecimalLiteral.Select(d => (object)d))
+                .Or(IntegerLiteral.Select(d => (object)d))
+                .Or(EmptyArrayLiteral)
```

Comment:
In cases where the next character can reliably determine which parser to choose, I'd lean more heavily on the `X` variants of combinators: `XMany` and `XOr`. They should give you equal or better performance and error location in most (maybe all) situations.

Reply:
Good suggestion, I'll try it out and put up another pr